pkg/controller/kubelet-config/kubelet_config_nodes.go

```diff
@@ -0,0 +1,434 @@
+		if err := retry.RetryOnConflict(updateBackoff, func() error {
+			var err error
+			if isNotFound {
+				_, err = ctrl.client.MachineconfigurationV1().MachineConfigs().Create(context.TODO(), mc, metav1.CreateOptions{})
+			} else {
+				_, err = ctrl.client.MachineconfigurationV1().MachineConfigs().Update(context.TODO(), mc, metav1.UpdateOptions{})
+			}
+			return err
+		}); err != nil {
+			return fmt.Errorf("Could not Create/Update MachineConfig, error: %v", err)
+		}
+		glog.Infof("Applied Node configuration %v on MachineConfigPool %v", key, pool.Name)
+		// fetch the kubeletconfigs
+		kcs, err := ctrl.mckLister.List(labels.Everything())
+		if err != nil {
+			return fmt.Errorf("could not get KubeletConfigs, err: %v", err)
+		}
+		for _, kc := range kcs {
+			// updating the existing kubeletconfigs with the updated nodeconfig
+			err := ctrl.syncKubeletConfig(kc.Name)
+			if err != nil {
+				return fmt.Errorf("could not update KubeletConfig %v, err: %v", kc, err)
+			}
+		}
+
+		// updating the nodeConfig status based on the machineconfigpool status
+		err = ctrl.updateNodeConfigStatus(pool, nodeConfig)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (ctrl *Controller) enqueueNodeConfig(nodeConfig *osev1.Node) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(nodeConfig)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", nodeConfig, err))
+		return
+	}
+	ctrl.nodeConfigQueue.Add(key)
+}
+
+func (ctrl *Controller) updateNodeConfig(old, cur interface{}) {
+	oldNode, ok := old.(*osev1.Node)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("Couldn't retrieve the old object from the Update Node Config event %#v", old))
+		return
+	}
+	newNode, ok := cur.(*osev1.Node)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("Couldn't retrieve the new object from the Update Node Config event %#v", cur))
+		return
+	}
+	if newNode.Name != ctrlcommon.ClusterNodeInstanceName {
+		message := fmt.Sprintf("The node.config.openshift.io \"%v\" is invalid: metadata.name Invalid value: \"%v\" : must be \"%v\"", newNode.Name, newNode.Name, ctrlcommon.ClusterNodeInstanceName)
+		glog.V(2).Infof(message)
+		ctrl.updateNodeConfigDegradedStatus(newNode, message, "UpdateProhibited")
+		return
+	}
+	if !reflect.DeepEqual(oldNode.Spec, newNode.Spec) {
+		// skipping the update in case of the Worker-Latency-Profile type transition from "Default" to "LowUpdateSlowReaction" and vice-versa
+		// (TODO) Ideally the user request has to be honoured, the transition need to be from Default -> Medium -> Low or vice-versa.
+		// Restricting the request for now until this process is automated in future.
+		if (oldNode.Spec.WorkerLatencyProfile == osev1.DefaultUpdateDefaultReaction && newNode.Spec.WorkerLatencyProfile == osev1.LowUpdateSlowReaction) || (oldNode.Spec.WorkerLatencyProfile == osev1.LowUpdateSlowReaction && newNode.Spec.WorkerLatencyProfile == osev1.DefaultUpdateDefaultReaction) {
```

Comment:
Hmm ok so this is the main check for whether the update is allowable or not.

So, I thought through the possibilities, and I think this will work so long as we only have 3 latency profiles.

i.e. this check is only for updates. So if I had 4 profiles that can only go from a->b->c->d and I want to skip b/c for example, I think can just do:
a->c (degrade, nothing happens)
c->d -> this actually works, because even though the actual config is on a, we are looking only at our object (on c) and desired (on d). So we essentially would have done a->d right? And this would let it through

In a more perfect scenario I think the controller creating the MC (sync function) should do the final check between current and desired. This should just enqueue and have the sync handle this.

I'm not against merging as is, but just for future, if we have more things to handle, the sync loop should be the place, and not individual informer checks.

Reply:
The idea here is to automate the process of this transition in future(as mentioned in the (TODO) comment) without rejecting the user request. That automatic transition needs a thorough design, testing and hence implemented a check here this way for now.